src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java

```diff
@@ -154,29 +143,38 @@ public UUID getPendingRepair()
         return ref.get().getPendingRepair();
     }
 
-    public int getManifestSize()
-    {
-        return manifest.components().size();
-    }
-
     @Override
     public void write(StreamSession session, DataOutputStreamPlus out, int version) throws IOException
     {
+        boolean streamEntirely = shouldStreamEntireSSTable && out instanceof AsyncStreamingOutputPlus;
```

Comment:
I believe this `out instaceof` is useless, if you check how that `out` is instantiated, it can never be something else.

Reply:
the API allows it to be other things, so would still require this check to avoid future changes causing class cast issues.

Reply:
@maedhroz @dcapwell but then, suddenly, we are perfecty happy to cast that into AsyncStreamingOutputPlus in without any checks and that write method also takes DataOutputStreamPlus ... Same for uncompressed / normal writer.
 https://github.com/apache/cassandra/blob/7542fd81abb445e10fbcc67929d742a9849b7129/src/java/org/apache/cassandra/db/streaming/CassandraCompressedStreamWriter.java#L61

Reply:
that seems problematic as `org.apache.cassandra.streaming.async.NettyStreamingMessageSender#sendControlMessage` calls this method (transitively) and passes in a `DataOutputBufferFixed`, which is not a `AsyncStreamingOutputPlus`.

I would need to look closer to see how they relate, but it does seem like this logic isn't safe.

Now, given the fact that this patch doesn't change this logic and the logic is already there; I am fine ignoring this fact in this JIRA, though that should be looked at closer as this seems like an edge case where `header.compressionInfo != null` might fail

Reply:
> org.apache.cassandra.streaming.async.NettyStreamingMessageSender#sendControlMessage

This is for `KeepAliveMessage`..

For `OutgoingStreamMessage`, it must be `AsyncStreamingOutputPlus `